[Feature] add kube-router (alternative CNI plugin that can also replace kube-proxy)

iwilltry42 opened the pull request:
This PR adds the functionality to deploy `kube-router` as a CNI-Plugin via RKE.
Additionally it allows one to disable the deployment of `kube-proxy` as the service proxy in general.

Please note, that this PR is probably still a little rough around the edges, since it's the first time that I touch those repositories. I guess that the handling of disabled kube proxy can be improved quite a lot (e.g. no need to generate the certs for it or handle it during reconciliation, etc.

Related PRs:
- https://github.com/rancher/kontainer-driver-metadata/pull/92
- ~https://github.com/rancher/types/pull/1058~ (dropped, because types are now included in rancher/rke directly)

Fixes: https://github.com/rancher/rke/issues/312 
Fixes: https://github.com/rancher/rke/issues/1432

iwilltry42 commented:
The merge conflicts can only be resolved after the other two PRs have been merged so that we can update the required go modules.

Further information: I tested the changes in this PR 3 times on two different set of nodes (Virtual Machines running Ubuntu 18.04.3 + Docker 18.09.9).

aderumier commented:
@iwilltry42 
Hi, thanks for your work !

I'll try to test it soon. 
I never build rke before, I'll try to do it.  (or maybe can you provide patched rke binary for testing ?)

iwilltry42 commented:
Hi @aderumier, a simple `make` will build the project using dapper (everything's inside containers).
If you want to modify the steps executed during the build, you can edit the file `scripts/ci`.
However to be able to build this PR, you also have to have the changes from the other two PRs locally + `replace` statements for the other two repos in the `go.mod` file of the RKE repo.

However, I've just uploaded the binary that I used for testing: https://github.com/iwilltry42/rke/releases/tag/20191213.4-kube-router

aderumier commented:
> Hi @aderumier, a simple `make` will build the project using dapper (everything's inside containers).
> If you want to modify the steps executed during the build, you can edit the file `scripts/ci`.
> However to be able to build this PR, you also have to have the changes from the other two PRs locally + `replace` statements for the other two repos in the `go.mod` file of the RKE repo.
> 
> However, I've just uploaded the binary that I used for testing: https://github.com/iwilltry42/rke/releases/tag/20191213.4-kube-router

Many thanks. I'm going to test it next week.

killcity commented:
Any updates on this merge?

x0day commented:
https://github.com/cloudnativelabs/kube-router/releases/tag/v1.0.0

kube-router has released 1.0.0, any updates?

iwilltry42 commented:
Updated this to work with the latest changes to the whole RKE architecture.
Changed some things in the integration tests (locally) to make them work with `kube-router`:
- install kubectl before executing RKE (since we're using the executable when kube-proxy is disabled)
- use an extended config in the tests:
```yaml
kubernetes_version: ${versions_to_test["${ver}"]}
nodes:
- address: rke-node-${node}
  role: [etcd, controlplane, worker]
  user: ubuntu
services:
  kubeproxy:
    enabled: false
network:
  plugin: kube-router
  options:
    kube_router_run_service_proxy: true
    kube_router_run_firewall: false
system_images:
  kube_router_cni: cloudnativelabs/kube-router:v1.0.1
```
This worked well for me :heavy_check_mark:

killcity commented:
Any idea when this will be approved?

mmack commented:
@superseb any news on this?

superseb commented:
I'm going to try to get this in for the release after the next release.

superseb commented:
Due to personal circumstances, this did not go into the latest release. I will try again the next release.

iwilltry42 commented:
@travisghansen 
> Will this allow disabling kube-proxy without installing kube-router? I’m after a sane solution to run with cilium and want to disable kube-proxy and manage cni myself entirely.

Yes it will :+1: 
See this section:
```yaml
services:
  kubeproxy:
    enabled: false
```

As personally, I'm now on rke2, I didn't push this PR forward anymore though so far.

superseb commented:
@iwilltry42 Are you willing to change the PR based on feedback or you want me to create a new PR for it?

iwilltry42 commented:
Hi @superseb , I just did a rebase onto `master`... seems like that was not a good idea given the new changeset that GitHub shows O.o

superseb commented on cluster/addons.go:
```diff
@@ -468,7 +468,8 @@ func (c *Cluster) deployWithKubectl(ctx context.Context, addonYaml string) error
 func (c *Cluster) doAddonDeploy(ctx context.Context, addonYaml, resourceName string, isCritical bool) error {
-	if c.UseKubectlDeploy {
+	if c.UseKubectlDeploy || (resourceName == NetworkPluginResourceName && !*c.Services.Kubeproxy.Enabled) {
+		// deploy via local kubectl if specified or if this is a network plugin and kubeproxy is disabled
```
This describes whats happening but not why, why the exception on this case

iwilltry42 replied:
I updated the comment.
IIRC it's because the in-cluster addon deploy jobs won't work if you don't have kube-proxy as the basic networking component active (or a replacement for it).

superseb commented:
I left some comments on this and KDM, if we can sync it up for v1.21 I can test it out

iwilltry42 commented:
@superseb addressed your comments :heavy_check_mark:

BloodyIron commented:
I can't tell which release of RKE has kube-router support, seemingly no release notes mention it that I can find... defeating the point of such things... and I really don't see a "nice" way to get kube-router going as the CNI in Rancher managed clusters vs Calico/Cilium/etc... so ... what's up with that?

BloodyIron commented:
@iwilltry42 was this ever actually merged into an RKE release that was published? I can't tell how on earth I actually use kube-router with rancher!

BloodyIron commented:
Yeah I don't see any of this code in any releases so far nor the master branch. It's as if it was just thrown away. What's going on here?!?!?! @superseb any ideas?

BloodyIron commented:
Well I guess I have to do this manually now considering Rancher devs are flat-out ignoring Kube-Router so far as I can tell, as well as this code @iwilltry42 wrote. Rancher is supposed to make this kind of stuff (selecting kube-router as the CNI) straightforward, but for reasons I can't find, or fathom, kube-router is just a plague to SUSE/Rancher devs, they don't want to touch it or it to be anywhere useful.

I need kube-router so that I can use it with MetalLB and get accurate Source IP info, since kube-proxy in cluster policy mode "Cluster" (forget the exact name) 100%, always, replaces the source IP with the node IP, which is completely asinine and useless for any of us trying to grant access, kick/ban people, read useful logs to help people, or anything you could ever want to do with knowing the right source IP for a user.

iwilltry42 commented:
Hey @BloodyIron , as far as I'm aware, this never made it into the code.
I updated it once for another try and since then didn't touch it.

As of now, I haven't used rke in years since I'm all in on K3s and RKE2 now.

BloodyIron commented:
@iwilltry42 aww okay, that's too bad. I need kube-router to get accurate Source-IP behind MetalLB Layer2 mode. Thanks for your work anyways! :)
